Add a new role editor

bl-nero opened the pull request:
So far, the editor only supports two fields, but it's a start. Not yet visible by default.

**To turn on the new UI, go to developer tools -> Application -> Local storage and add a `grv_teleport_use_new_role_editor` key set to `true`.** This will be lifted once UI is good to be released.

Known issue: switching back to Standard after displaying the YAML editor panel always warns about non-standard fields, as this is a bare-bones version that doesn't even understand the default flags.

[Figma designs](https://www.figma.com/design/v6GunK50D2VC7w7I2FBDNf/Access-(Management)?node-id=4956-32313&node-type=canvas&m=dev)

![Screenshot 2024-10-02 at 19 03 44](https://github.com/user-attachments/assets/a7d07e5f-ebe8-4d6f-bca6-de52cdc15226)

![Screenshot 2024-10-02 at 19 04 25](https://github.com/user-attachments/assets/54c31278-d702-4d18-b820-46b8e5382936)

bl-nero committed:
Add a new role editor

So far, the editor only supports two fields, but it's a start.

avatus reviewed (approved):
![Screenshot 2024-10-02 at 3 13 28 PM](https://github.com/user-attachments/assets/6cb67da9-baf1-4363-b99f-768d24bb667d)
this indicator when opening the editor seems to be off center.

pretty cool stuff! can't wait to see this grow

kimlisa reviewed (approved):
non blocker but the routing rule editor loading/error handling is kinda opposite to how it works here, and it was to let re-tries feel more intuitive (IMO)

for routing rule editor on change tab, the user still stays in the the current state but disabled and loading, and only on success go to the clicked tab, this way on failure, user is still on the last success state and the user can try again by clicking on the tab again. 

i liked this approach b/c it feels less jarring but i don't remember taking it to the design team so perhaps there's UX issues i'm not seeing. i've added it to the design topic for next tuesday so that we can keep the editor behaviors consistent

if this was already discussed, well guess we gotta change the routing editor behavior 😅

bl-nero commented on web/packages/teleport/src/Roles/RoleEditor/RequiresResetToStandard.tsx:
```diff
@@ -0,0 +1,32 @@
+
+export const RequiresResetToStandard = ({ reset }: { reset(): void }) => (
+  <Info>
+    <Text>
+      Some fields were not readable by the standard editor. To continue editing,
+      go back to YAML editor or reset the affected fields to standard settings.
+    </Text>
```
@avatus That was one of my (more ambitious) plans, but I'm not sure if I'll make it before v17. It's definitely a good idea in case of this editor, as it's gonna be quite complex.

bl-nero commented:
> non blocker but the routing rule editor loading/error handling is kinda opposite to how it works here, and it was to let re-tries feel more intuitive (IMO)
> 
> for routing rule editor on change tab, the user still stays in the the current state but disabled and loading, and only on success go to the clicked tab, this way on failure, user is still on the last success state and the user can try again by clicking on the tab again.
> 
> i liked this approach b/c it feels less jarring but i don't remember taking it to the design team so perhaps there's UX issues i'm not seeing. i've added it to the design topic for next tuesday so that we can keep the editor behaviors consistent
> 

@kimlisa I see your point, and I think I somehow missed it while adapting the editor to the role domain, and then saw some inconsistency, and "solved" it this way. I think you're right here, I'll try to change it.

bl-nero commented:
> this indicator when opening the editor seems to be off center.

@avatus Oops, forgot to center it. :)

bl-nero commented:
@zmb3 Zac, looks like this change will require an admin approval due to its size.

bl-nero committed:
Review

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor

bl-nero commented:
@kimlisa Addressed your feedback in https://github.com/gravitational/teleport/pull/47100/commits/271d433d9d96a88f8f6460ef523bf9665a7c1ade. It simplified some things. Can you verify that this is what you meant?

kimlisa commented:
> @kimlisa Addressed your feedback in [271d433](https://github.com/gravitational/teleport/commit/271d433d9d96a88f8f6460ef523bf9665a7c1ade). It simplified some things. Can you verify that this is what you meant?

yeah that looks great 👍 , thanks

bl-nero committed:
Center the loading indicator
